nexus/db-model/src/instance.rs

```diff
@@ -220,6 +246,32 @@ impl InstanceRuntimeState {
             dst_propolis_id: None,
             migration_id: None,
             gen: Generation::new(),
+            time_last_auto_restarted: None,
+        }
+    }
+
+    /// Returns `true` if this instance can be automatically restarted by the
+    /// provided [`InstanceAutoRestart`] policy.
+    pub fn can_reincarnate(&self, policy: InstanceAutoRestart) -> bool {
+        // Instances only need to be automatically restarted if they are in the
+        // `Failed` state.
+        if self.nexus_state != InstanceState::Failed {
+            return false;
+        }
+
+        // Check if the instance's configured auto-restart policy permits the
+        // control plane to automatically restart it.
+        match policy {
+            InstanceAutoRestart::Never => false,
+            InstanceAutoRestart::AllFailures => true,
+            // TODO(eliza): currently, we don't have the ability to determine
+            // whether an instance is failed because the sled it was on has
+            // rebooted, or because the individual Propolis VMM crashed. For
+            // now, we assume all failures are VMM failures rather than sled
+            // failures. In the future, we will need to determine if a failure
+            // was a sled-level or VMM-level failure, and use that here to
+            // determine whether or not the instance is restartable.
```

Comment:
> What's the purpose of the SledFailuresOnly policy? i.e., when would users use this?

My thought process here was that there are basically a few broad categories of reasons that an instance might have failed:

1. **Propolis panicked.** This may be due to some guest-specific misbehavior, or a Propolis bug that is triggered by something specific about that instance. If an instance tries to do something that makes Propolis crash (whether or not that's misbehavior on the guest's part, or on Propolis' part), it's plausible that the guest might do the same thing again. So, you might reasonably want to require human intervention to determine if that instance ought to be restarted before just going and doing it.
2. **The sled failed and was expunged.** The instances on it were just unlucky enough to have been left behind on a sled that had a hardware fault. In that case, it's definitely not those instances' fault that they've failed, and you'd almost certainly want them restarted elsewhere (unless there are no longer sufficient rack-level resources for them).
3. **The sled unexpectedly rebooted.** Similarly to case (2), it's much less likely that this is the _instance's_ fault than case (1), so you probably want the instances to come back. And, it's less likely that there will no longer be sufficient capacity for those instances, because all the sleds are still present.

Therefore, because an individual Propolis panicking (case 1) seems much likelier to be the guest VM's fault than the cases where the whole sled had some hardware problem (2 and 3), it seemed to me that it would be desirable for a user to be able to select a restart policy that only restarts the instances in cases where the fault is known to effect the entire sled. In particular, that seems potentially desirable as a default restart policy when instances haven't been explicitly configured to always auto-restart.

Regarding the second question,

> So if we're going to have a policy that we don't implement, if someone has opted into restarts, I think I'd rather say "sorry, we restarted your instance even though it wasn't a sled failure because we can't tell yet" than "sorry, we didn't restart your instance even though it was a sled failure because we can't tell yet". Although I also wonder if we should just not expose this policy externally until we implement it **exactly?**

My plan is to implement this policy more exactly in a subsequent branch, which I've already started working on (#6559). I wanted to do that in a separate PR in the hopes of making the review a bit less daunting, and so I could start getting feedback on the general approach to instance reincarnation sooner. Right now, none of this stuff is exposed to the user: the web UI doesn't have any way to select restart policies yet, and the CLI doesn't have an argument for it --- the only way to select any of these policies at present is to manually construct JSON and POST it at Nexus, [like I did here when testing this](https://github.com/oxidecomputer/omicron/pull/6503#issuecomment-2342107307). My intention was that this branch and #6559 should be in the same release, and that we wouldn't expose the `SledFailuresOnly` policy in any user-facing interface until it actually does something.

Does that seem reasonable to you?

Reply:
> If an instance tries to do something that makes Propolis crash (whether or not that's misbehavior on the guest's part, or on Propolis' part), it's plausible that the guest might do the same thing again. So, you might reasonably want to require human intervention to determine if that instance ought to be restarted before just going and doing it.

This is kind of a half-baked thought, but I'm starting to wonder if users should be responsible for reasoning about this distinction. It seems like we're trying to emphasize this difference:

- If an instance's VMM is lost because its sled rebooted, it's relatively likely (though not absolutely certain!) that the instance/VMM didn't cause the problem, and thus is relatively more likely that the instance will run happily if it's restarted on another sled. So it makes sense to restart instances more aggressively when they fail because their sleds bounce.
- But if an instance's VMM panics, it's much more likely (compared to a sled reboot) that the problem has something to do with that instance or its guest (the guest did something we didn't handle properly, the instance is configured in a way that triggers a Propolis bug deterministically, etc.). So it's relatively less likely that restarting an instance will make everything happy again, and so we should be less aggressive about restarting instances that are panicking individually.

Is this distinction something a *user* should have to care about? The more I think about it the less sure I get. It seems like it might be much simpler to say that if auto-restart is set, the control plane will at least *attempt* to restart a Failed instance, but it reserves the right to stop trying at any time, for any reason. Then we'd just implement some policy that throttles restart attempts much faster for instance-specific failures than sled-wide failures.

I dunno. Like I said: half-baked thought, not necessarily a request to change anything in this PR; but the more I think about how we'd explain the difference between these policies (and why you might choose one over the other), the more I wonder if we should just make it all Nexus's problem.

Reply:
> My thought process here was that there are basically a few broad categories of reasons that an instance might have failed:

I'm not sure there's that much difference between Propolis panicking and a sled panicking or rebooting.  Either of these may _or_ may not have been caused by the instance itself.  I'm not sure we can attribute a probability that the instance caused it in either case.  If we don't know what the cause was, I'd be surprised if a user knew or cared about this distinction.  Propolis, the host OS, and the sled are all part of the underlying infrastructure as far as they're concerned.  I'm definitely open to other views on this!

There's arguably a separate question of: if users don't already have strong expectations here, which would we prefer?  At Joyent, as operators, it was not uncommon to want to reboot a server for various reasons.  In a world where customers build applications assuming the infrastructure can restart VMs at any time (because servers do fail, after all, but we can restart the instances), this can give operators a lot of flexibility.  Conversely, if we guide customers towards choices that will mean their application goes offline when sleds reboot, I think we might be setting ourselves up for making panics and upgrades and other maintenance operations _much_ more disruptive than they might otherwise be.

---

On the second question: I do think it seems reasonable to separate these especially if you've got fairly high confidence the second will land in this release.  I also think having it in the API counts as exposing it to users.  I generally find this to be a tradeoff: in principle it's nice to have a placeholder variant that gets implemented later.  In the meantime though the code is in kind of a weird state, and sometimes we get busy with other things and don't come back to fix it as soon as we thought.  Or honestly what happens more often for me is that by the time I go implement the behavior later, I find that the way I tried to make it easier in the first PR (e.g., with an enum variant) isn't right and needs to change anyway.  But at the same time, if it makes the follow-on PR simpler, that's useful too!

Reply:
I will admit that, at this point, I am beginning to wonder if having these policies even makes sense. @gjcolombo's argument in https://github.com/oxidecomputer/omicron/pull/6503#discussion_r1757739975 is compelling --- perhaps the real use of distinguishing whether a failure was sled-wide or instance-specific is to enforce two separate cooldowns for different classes of failures, or something like that, rather than allowing the user to say "I only want to restart instances on specific failure classes". So, maybe we should really just change the policy enum back to just "always"/"never", with the option to add policies like exponential backoffs or a restart budget as additional policies later.

I think that failure reason tracking is still important, though. At the very least, we might want to not count an operator-initiated sled expungement (or even an operator-initiated sled restart for something like a mupdate) against the cooldown period, because those aren't actually signs that the instance might have done something naughty...

Reply:
Okay, I've reworked the auto-restart policies in commit https://github.com/oxidecomputer/omicron/pull/6503/commits/b4183b1ab2616c86e0d7ec5f46e2ac99e959ee04. Now, rather than specifying the _classes_ of failure that may be restarted, they specify a desired "quality of service" for instance restarting. Presently, the only options are `Never` (don't automatically restart this under any circumstances) and `BestEffort` (Nexus does "whatever it thinks is best, which may include cooldowns, backoff periods, choosing not to restart an instance under some circumstances, etc, but is not *guaranteed* to get any specific behavior). 

In the future, we may want to add additional policies like "always do an exponential backoff", "guaranteed immediate auto-restarts at all times", or "enforce the cooldown on consecutive failures of the same kind", but implementing these ought to be driven by user requirements for specific restart policies. For now, the `BestEffort` policy is intended as a generic "make reasonable attempts to keep this instance running if possible" option that should probably be the default for most uses.